streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -0,0 +1,398 @@
+
+    private volatile boolean open = false;
+
+    InMemoryWindowStore(final String name,
+                               final Serde<K> keySerde,
+                               final Serde<V> valueSerde,
+                               final long retentionPeriod,
+                               final long windowSize,
+                               final boolean retainDuplicates,
+                               final String metricScope) {
+        this.name = name;
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.retentionPeriod = retentionPeriod;
+        this.windowSize = windowSize;
+        this.retainDuplicates = retainDuplicates;
+        this.metricScope = metricScope;
+
+        this.segmentMap = new TreeMap<>();
+    }
+
+    @Override
+    public String name() {
+        return this.name;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void init(final ProcessorContext context, final StateStore root) {
+        this.context = (InternalProcessorContext) context;
+
+        // construct the serde
+        this.serdes = new StateSerdes<>(
+            ProcessorStateManager.storeChangelogTopic(context.applicationId(), name),
+            keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
+            valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+
+        final StreamsMetricsImpl metrics = this.context.metrics();
+        final String taskName = context.taskId().toString();
+        expiredRecordSensor = metrics.storeLevelSensor(
+            taskName,
+            name(),
+            "expired-window-record-drop",
+            Sensor.RecordingLevel.INFO
+        );
+        addInvocationRateAndCount(
+            expiredRecordSensor,
+            "stream-" + metricScope + "-metrics",
+            metrics.tagMap("task-id", taskName, metricScope + "-id", name()),
+            "expired-window-record-drop"
+        );
+
+        if (root != null) {
+            context.register(root, (key, value) -> {
+                put(extractStoreKey(key, serdes), serdes.valueFrom(value), extractStoreTimestamp(key));
+            });
+        }
+        this.open = true;
+    }
+
+    @Override
+    public void put(final K key, final V value) {
+        put(key, value, context.timestamp());
+    }
+
+    @Override
+    public void put(final K key, final V value, final long windowStartTimestamp) {
+        removeExpiredSegments();
+        maybeUpdateSeqnumForDups();
+
+        if (windowStartTimestamp <= this.context.streamTime() - this.retentionPeriod) {
+            expiredRecordSensor.record();
+            LOG.debug("Skipping record for expired segment.");
+        } else {
+            if (value != null) {
+                this.segmentMap.computeIfAbsent(windowStartTimestamp, t -> new TreeMap<>());
+                this.segmentMap.get(windowStartTimestamp).put(new WrappedK<>(key, seqnum), value);
```

Comment:
If retainDuplicates is false, could we not storing the sequence suffix at all to save 4 bytes per record?

Reply:
I think this will be more straightforward after we remove the generics since we can just tack on the seqnum Bytes or not

Reply:
Sounds good.